driver/src/main/com/mongodb/GridFSBucketImpl.java

```diff
@@ -0,0 +1,316 @@
+        checkCreateIndex();
+        return new GridFSUploadStreamImpl(filesCollection, chunksCollection, new ObjectId(), filename, chunkSize, metadataBsonDocument);
+    }
+
+    @Override
+    public ObjectId uploadFromStream(final String filename, final InputStream source) {
+        return uploadFromStream(filename, source, new GridFSUploadOptions());
+    }
+
+    @Override
+    public ObjectId uploadFromStream(final String filename, final InputStream source, final GridFSUploadOptions options) {
+        GridFSUploadStream uploadStream = openUploadStream(filename, options);
+        int chunkSize = options.getChunkSizeBytes() == null ? chunkSizeBytes : options.getChunkSizeBytes();
+        byte[] buffer = new byte[chunkSize];
+        int len;
+        try {
+            while ((len = source.read(buffer)) != -1) {
+                uploadStream.write(buffer, 0, len);
+            }
+            uploadStream.close();
+        } catch (IOException e) {
+            throw new MongoGridFSException("IO Exception when reading from the InputStream", e);
+        }
+        return uploadStream.getFileId();
+    }
+
+    @Override
+    public GridFSDownloadStream openDownloadStream(final ObjectId id) {
+        BsonDocument fileInfo = filesCollection.find(new BsonDocument("_id", new BsonObjectId(id))).first();
+        if (fileInfo == null) {
+            throw new MongoGridFSException(format("No file found with the ObjectId: %s", id));
+        }
+        return new GridFSDownloadStreamImpl(fileInfo, chunksCollection);
+    }
+
+    @Override
+    public void downloadToStream(final ObjectId id, final OutputStream destination) {
+        downloadToStream(openDownloadStream(id), destination);
+    }
+
+    @Override
+    public GridFSDownloadStream openDownloadStreamByName(final String filename) {
+        return openDownloadStreamByName(filename, new GridFSDownloadByNameOptions());
+    }
+
+    @Override
+    public GridFSDownloadStream openDownloadStreamByName(final String filename, final GridFSDownloadByNameOptions options) {
+        return new GridFSDownloadStreamImpl(getFileByName(filename, options), chunksCollection);
+    }
+
+    @Override
+    public void downloadToStreamByName(final String filename, final OutputStream destination) {
+        downloadToStreamByName(filename, destination, new GridFSDownloadByNameOptions());
+    }
+
+    @Override
+    public void downloadToStreamByName(final String filename, final OutputStream destination, final GridFSDownloadByNameOptions options) {
+        downloadToStream(openDownloadStreamByName(filename, options), destination);
+    }
+
+    @Override
+    public GridFSFindIterable<Document> find() {
+        return new GridFSFindIterableImpl<Document>(getFilesCollection().find());
+    }
+
+    @Override
+    public <TResult> GridFSFindIterable<TResult> find(final Class<TResult> resultClass) {
```

Comment:
It is in the Spec - with the explicit options for the find.

It also mirrors all our `MongoIterable` so the usability is consistent.